src/tqec/compile/specs/base.py

```diff
@@ -70,3 +67,63 @@ def __call__(self, spec: CubeSpec) -> CompiledBlock:
             a `CompiledBlock` based on the provided `CubeSpec`.
         """
         ...
+
+
+@dataclass(frozen=True)
+class PipeSpec:
+    """Specification of a pipe in a block graph.
+
+    The `PipeSpec` is used to determine the substitution rules between the two
+    `CompiledBlock`s connected by the pipe. The substitution rules are used to
+    update the layers of the `CompiledBlock`s based on the plaquettes in the
+    `Substitution`.
+
+    Attributes:
+        spec1: the cube specification of the first cube. By convention, the cube
+            corresponding to `spec1` should have a smaller position than the cube
+            corresponding to `spec2`.
+        spec2: the cube specification of the second cube.
+        pipe_type: the type of the pipe connecting the two cubes.
+    """
+
+    spec1: CubeSpec
+    spec2: CubeSpec
+    pipe_type: PipeType
+
+
+@dataclass(frozen=True)
+class Substitution:
+    """Collection of plaquettes categorized by the layer index.
+
+    This specifies how to substitute plaquettes in the two `CompiledBlock`s
+    connected by a pipe. When applying the substitution, the plaquettes in
+    the map will be used to update the corresponding layer in the `CompiledBlock`.
+
+    Both the source and destination maps are indexed by the layer index in the
+    `CompiledBlock`. The index can be negative, which means the layer is counted
+    from the end of the layers list.
+
+    Attributes:
+        src: a mapping from the index of the layer in the source `CompiledBlock` to
+            the plaquettes that should be used to update the layer.
+        dst: a mapping from the index of the layer in the destination `CompiledBlock`
+            to the plaquettes that should be used to update
+    """
+
+    src: Mapping[int, Plaquettes]
+    dst: Mapping[int, Plaquettes]
```

Comment:
I think that the convention in Python for typing is the following:
- "abstract" or "generic" types (such as `Mapping`) are used in places that are visible to users,
- concrete types are used everywhere else.

Is there any advantage in having `Mapping` instead of a regular `FrozenDefaultDict`/`dict`/`defaultdict` here?

Reply:
I think no. I will change it to `dict`.